core/sr-primitives/src/weights.rs

```diff
@@ -56,6 +64,26 @@ pub enum TransactionWeight {
 	Free,
 }
 
+/// A wrapper for fee multiplier.
+/// This is to simulate a `Perbill` which is greater than `1`.
+///
+/// The fee multiplier is always multiplied by the weight (as denoted by `TransactionWeight` on a
+/// per-transaction basis with `#[weight]` annotation) of the transaction to obtain the final fee.
+///
+/// One can define how this conversion evolves based on the previous block weight by implementing
+/// the `FeeMultiplierUpdate` type of the `system` trait.
+#[cfg_attr(feature = "std", derive(PartialEq, Eq, Clone, Debug))]
+#[derive(Encode, Decode)]
+pub enum FeeMultiplier {
+	/// Should be interpreted as a positive ratio added to the weight, i.e. `weight + weight * p`
+	/// where `p` is a small `Perbill`.
+	///
+	Positive(Perbill),
```

Comment:
This will only work up until `u32::max_value()/1000000000`x, after which we won't be able to encode any higher values. Is this not a problem?

Reply:
strange, why is that? Like some kind of a codec/storage limitation? 

Yes, it will be a problem for sure. These perbill numbers are most often small, because of the formula being `1 + something_small - something_small`, but we don't have a guarantee.

If there is no way to store a Perbill safely, I can also store a raw `u32` and treat it as `Perbill` only before multiplication.

Reply:
I don't think the range is correct.

`fm = 1 + (diff) + (diff)^2 / 2` and `fee = weight * fm`. 

the power of two of diff is super small; most often collapsed to zero with Perbill so that's not the dominant factor + it will have the same logic is below:

 the diff itself is the subtraction of two saturation levels (`s - s*`) (`s*` hardcoded to 0.25 atm) and the other being the current saturation (again being in range `[0,1]`). Hence `diff is_in_range [1-s*, -s*]`. So the positive side is safe + we need the enum wrapper to represent the case where the value is a _negative perbill_

Reply:
hmm I think in this implementation it is more like: `fm = (diff) + (diff)^2 / 2` and `fee = weight * (1 + fm)`

but I had a different view because I considered fm as a slow long term adjustment thus it could potentially slowly grow to 2 or 3 or whatever values.

Reply:
> Like some kind of a codec/storage limitation?

It's simply that `Perbill` is a `u32` internally, which is scaled by 1/1,000,000. as such the maximum value it can encode us `u32::max_value / 1000000` =~ 4.

If this needs to potentially grow to factors over 4, then it should be done through another type.

Reply:
Currently Perbill internal u32 is always capped to one billion. And I might have used this property.

Reply:
I think it should stay like that. A Per-thing is always [0, 1]. For now, this PR can proceed since the `FeeMultiplier` is simulating a fixed point number in the range [-1, infinity] (tuple of u32 and Perbill) which is enough. I will also work on a separate PR to add a `Fixed64` type which does the same in a generic way.